File: src/main/java/com/github/project_njust/ccf_manager/sql/impl/ExamInfoManagerimpl.java
package com.github.project_njust.ccf_manager.sql.impl;

import com.github.project_njust.ccf_manager.SQLManager;
import com.github.project_njust.ccf_manager.model.ExamInfo;
import com.github.project_njust.ccf_manager.sql.IExamInfoManager;

import java.sql.*;

public class ExamInfoManagerimpl implements IExamInfoManager {
    @Override
    public int insertExamInformation(ExamInfo record) {
        Connection conn= SQLManager.getConnection();
        int result = 0;
        PreparedStatement pst = null;
        try {
            String sql = "insert into examinformation(ExamScore, ExamDate, Data)  value(?,?, '{}')";
            pst = (PreparedStatement) conn.prepareStatement(sql);
            pst.setInt(1,record.getExamscore());
            pst.setDate(2, (Date) record.getExamdate());
            result = pst.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        SQLManager.evictConnection(conn);
        return result;
    }

    @Override
    public int updateExamInformation(ExamInfo record, int ExamID) {
        Connection conn= SQLManager.getConnection();
        int result=0;
        PreparedStatement pst = null;
        try {
            String sql = "update examinformation set ExamScore=?,ExamDate=?,Data=? where ExamID=?";
            pst = (PreparedStatement) conn.prepareStatement(sql);
            pst.setInt(1,record.getExamscore());
            pst.setDate(2, (Date) record.getExamdate());
//TODO data
            pst.setInt(4,record.getExamid());
            result = pst.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        SQLManager.evictConnection(conn);
        return result;
    }

    @Override
    public int deleteExamInformation(int ExamId) {
        Connection conn= SQLManager.getConnection();
        int result=0;
        PreparedStatement pst = null;
        try {
            String sql = "delete from examinformation where ExamID = ?";
            pst = (PreparedStatement) conn.prepareStatement(sql);
            pst.setInt(1,ExamId);
            result = pst.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        SQLManager.evictConnection(conn);
        return result;
    }

    @Override
    public ExamInfo selectExamInformationByExamId(int ExamId) {
        Connection conn= SQLManager.getConnection();
        PreparedStatement pst = null;
        ResultSet result=null;
        try {
            String sql = "select * from examinformation where ExamId= ?";
            pst = (PreparedStatement) conn.prepareStatement(sql);
            pst.setInt(1,ExamId);
            result = pst.executeQuery();
            if(result.next()){
                return new ExamInfo(result.getInt("ExamID"), result.getInt("examscore"),result.getDate("examsdate"));
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        SQLManager.evictConnection(conn);
        return null;
    }
}

File: src/main/java/com/github/project_njust/ccf_manager/sql/impl/ExamInfoManagerImpl.kt
package com.github.project_njust.ccf_manager.sql.impl

import com.github.project_njust.ccf_manager.SQLManager
import com.github.project_njust.ccf_manager.model.ExamInfo
import com.github.project_njust.ccf_manager.sql.IExamInfoManager
import com.github.project_njust.ccf_manager.wrapper.json.JsonSection
import kotlinx.coroutines.runBlocking
import java.sql.Date

object ExamInfoManagerImpl : IExamInfoManager {
    override fun createExamInformation(score: Int, date: Date): ExamInfo = runBlocking {
        SQLManager.coroutinesConnection {
            val ps = this.prepareStatement("INSERT INTO ExamInformation (EXAMSCORE, EXAMDATE, DATA) VALUES (?, ?, ?)")
            ps.setInt(1, score)
            ps.setDate(2, date)
            ps.setString(3, "{}")
            ps.executeUpdate()
        }
        return@runBlocking getLatestInfo()
    }

    override fun getLastInfo(): ExamInfo? {
        SQLManager.operateConnection {
            val ps = this.prepareStatement("SELECT * FROM ExamInformation WHERE EXAMID IN (SELECT MAX(EXAMID) FROM ExamInformation)")
            val rs = ps.executeQuery()
            if (rs.next()) {
                return ExamInfo(
                        rs.getInt(1),
                        rs.getInt(2),
                        rs.getDate(3),
                        rs.getString(4),
                        JsonSection.readFromJson(rs.getString(5))
                )
            }
        }
        return null
    }

    private suspend fun getLatestInfo(): ExamInfo {
        return SQLManager.asyncDeferred {
            val ps = this.prepareStatement("SELECT * FROM ExamInformation WHERE EXAMID IN (SELECT MAX(EXAMID) FROM ExamInformation)")
            val rs = ps.executeQuery()
            if (rs.next()) {
                ExamInfo(
                        rs.getInt(1),
                        rs.getInt(2),
                        rs.getDate(3),
                        rs.getString(4),
                        JsonSection.readFromJson(rs.getString(5))
                )
            } else {
                throw IllegalStateException("查找不到考试信息")
            }
        }.await()!!
    }

    override fun updateExamInformation(ei: ExamInfo) {
        SQLManager.operateConnection {
            val ps = this.prepareStatement("UPDATE ExamInformation SET EXAMSCORE = ?, EXAMDATE = ?, DATA = ?, CODE = ? WHERE EXAMID = ?")
            ps.setInt(1, ei.examscore)
            ps.setDate(2, ei.examdate)
            ps.setString(3, ei.data.toString())
            ps.setString(4, ei.code)
            ps.setInt(5, ei.examid)
            ps.executeUpdate()
        }
    }

    override fun selectExamInformationByExamId(ExamId: Int): ExamInfo? {
        SQLManager.operateConnection {
            val ps = this.prepareStatement("SELECT * FROM ExamInformation WHERE EXAMID = ?")
            ps.setInt(1, ExamId)
            val rs = ps.executeQuery()
            if (rs.next()) {
                return ExamInfo(ExamId, rs.getInt(2), rs.getDate(3), rs.getString(4), JsonSection.readFromJson(rs.getString(5)))
            }
        }
        return null
    }
}